pkg/services/ngalert/state/state_tracker.go

```diff
@@ -0,0 +1,90 @@
+package state
+
+import (
+	"fmt"
+	"github.com/grafana/grafana-plugin-sdk-go/data"
+	"github.com/grafana/grafana/pkg/services/ngalert/eval"
+	"github.com/grafana/grafana/pkg/services/ngalert/models"
+	"sync"
+)
+
+type AlertState struct {
+	UID     string
+	CacheId string
+	Labels  data.Labels
+	State   eval.State
+	Results []eval.State
+}
+
+type cache struct {
+	cacheMap map[string]AlertState
+	mu       sync.Mutex
+}
+
+var stateCache cache
+
+func Init() {
+	stateCache = cache{
+		cacheMap: make(map[string]AlertState, 0),
+		mu:       sync.Mutex{},
+	}
+}
+
+func (c *cache) getOrCreate(uid string, result eval.Result) AlertState {
+	c.mu.Lock()
+	defer c.mu.Unlock()
+
+	idString := fmt.Sprintf("%s %s", uid, result.Instance.String())
+	if state, ok := c.cacheMap[idString]; ok {
+		return state
+	} else {
+		state := AlertState{
+			UID:     uid,
+			CacheId: idString,
+			Labels:  result.Instance,
+			State:   result.State,
+			Results: []eval.State{result.State},
+		}
+		c.cacheMap[idString] = state
+		return state
+	}
+}
+
+func (c *cache) update(stateEntry AlertState) {
+	c.mu.Lock()
+	defer c.mu.Unlock()
+	c.cacheMap[stateEntry.CacheId] = stateEntry
+}
+
+func (c *cache) getStateForEntry(stateId string) eval.State {
+	c.mu.Lock()
+	defer c.mu.Unlock()
+	return c.cacheMap[stateId].State
+}
+
+func ProcessEvalResults(uid string, results eval.Results, condition models.Condition) {
```

Comment:
If this is the top-level function we'll use to determine to process, can we get the firing alerts as a return value?

Reply:
Absolutely! I'll make that change in a future PR though since the functionality to iterate of the map is not yet in here.

Reply:
Sounds good to me! Keep in mind though that we don't need to list as part of the evaluation -> notification workflow. The return value I'm referring to here is so that we can transform it into the payload the "Alertmanager" expects.